### PR TITLE
paper review

### PR DESCRIPTION
**Pull request by ayush-1506**
Should be easier to review it here.

cc @ablaom

### CODE REVIEW

**Comment by ablaom** [comment]
Got it. Will look at this later today.

**Comment by ayush-1506** [comment]
ps This is identical to the final draft I sent you, haven't made any changes yet.

File: paper/paper.md [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+   index: 1
+ - name: Department of Computer Science, University of Auckland
+   index: 2
+date: 26 March 2021
+bibliography: paper.bib
+---
+
+# Introduction
+
+We present _MLJFlux.jl_ [@MLJFlux.jl], an interface between the _MLJ_ [@Blaom2020] machine learning toolbox and _Flux.jl_ [@Innes2018] deep learning framework written in the _Julia_ [@Julia-2017] programming language. MLJFlux makes it possible to implement supervised deep learning models while adhering to the MLJ workflow. This means that users familiar with the MLJ design can write their models in Flux with a few slight modifications and perform all tasks provided by the MLJ model spec which includes evaluation, fitting on a dataset, hyperparameter tuning or plotting. The interface also provides options to train the model on different hardware and warm start the model after changing some specific hyper-parameters.
```
Comment (ablaom) [inline paper/paper.md:24]: The citations should not interrupt the flow of the sentence. So, for instance, 

"... between the MLJ [@Julia-2017] machine learning toolbox and ..." ✔️ 

"... between the MLJ machine learning toolbox [@Julia-2017] and ..."  ❌ 

This applies to the other 2 citations in this paragraph

File: paper/paper.md [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+
+Julia aims to solve the _"two language problem"_ in scientific computing, where high-level languages such as Python or Ruby are easy to use but often slow and low-level languages are fast but difficult to use. Using _just-in-time_ compilation and _multiple dispatch_, Julia makes the best of both worlds by matching the performance of low-level languages [@JuliaBenchmarks] while also being adaptable.
```
Comment (ablaom) [inline paper/paper.md:26]: "Julia solves the ...." not "Julia aims to ...."

File: paper/paper.md [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+
+# Statement of Need
+
+While MLJ supports multiple statistical models, it lacks support for any kind of deep learning model. MLJFlux adds support for this by interfacing MLJ with the Flux.jl deep learning framework. Converting a Flux model into an MLJ spec can be done by wrapping it in the appropriate MLJFlux container and specifying other hyper-parameters such as the loss function, optimizer, epochs, and so on. This can now be used like any other MLJ model. It can be further be wrapped in an MLJ _machine_, which in MLJ terminology is a way of binding a container for hyper-parameters (i.e. MLJ _model_) with data, and then any functionalities that work with MLJ machines can be invoked. MLJFlux also gives the ability to warm start the training upon changing certain model hyper-parameters.
```
Comment (ablaom) [inline paper/paper.md:32]: Suggest we delete "It can be further wrapped ... can be invoked" (see mm email).

Suggest replacing last sentence in this paragraph with:

 "MLJFlux models implement the MLJ warm restart interface, which means training can be restarted from where it left off, when the number of epochs is increased or the optimiser settings (e.g., learning rate) are modified. Consequently, an MLJFlux model can also be wrapped as an MLJ `IteratedModel`, making early stopping, model snapshots, callbacks, cyclic learning rates, and other controls available. "

File: paper/paper.md [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+
+MLJFlux provides four containers that we can enclose our Flux model in. Each model is derived from either `MLJModelInterface.Probabilistic` or `MLJModelInterface.Deterministic` to follow the MLJ design, depending on the type of task (regression or prediction). At the core of each wrapper is a _builder_ attribute that specifies the Flux model. The _builder_ defines instructions on how to build the Flux model once we've seen the data.
```
Comment (ablaom) [inline paper/paper.md:34]: "(regression or classification)" and not "(regression and prediction)".

Not sure this first sentence is absolutely necessary, although perhaps it *is* something we could make clearer in the readme documentation. If you think it helps, do leave it here. 

Can probably combine the last two sentences into one. Also "Flux model" is a bit vague if you are not familiar with Flux. As you know, the word "model" means different things to different people in different contexts. Maybe  "neural network architecture" is better here. So, perhaps something like:

"At the core of each wrapper is a _builder_ attribute that specifies the neural network architecture (Flux model) given the shape of the data. "

**Review by ablaom — COMMENTED** [review]
Finally, there are the comments I made about the plotting in the email. Happy for you to make the call about this.

File: paper/paper.md [hunk 5]
```diff
@@ -21,17 +21,17 @@ bibliography: paper.bib
...
+We present _MLJFlux.jl_ [@MLJFlux.jl], an interface between the _MLJ_ machine learning toolbox [@Blaom2020] and _Flux.jl_ deep learning framework [@Innes2018] written in the _Julia_ programming language[@Julia-2017]. MLJFlux makes it possible to implement supervised deep learning models while adhering to the MLJ workflow. This means that users familiar with the MLJ design can write their models in Flux with a few slight modifications and perform all tasks provided by the MLJ model spec. The interface also provides options to train the model on different hardware and warm start the model after changing some specific hyper-parameters.
```
Comment (ablaom) [inline paper/paper.md:24]: missing "the" before "_Flux.jl_"

**Review by ablaom — CHANGES_REQUESTED** [review]
Just that missing "the". 

Thanks for this.

**Comment by ayush-1506** [comment]
@ablaom Done. Please let me know if you have any final thoughts.

**Comment by ablaom** [comment]
No. I didn't really check the citations carefully, but JOSS has that automated and you can change them later. Let's proceed with a submission and see what feedback we get. 

Again many thanks for leading this!

**Review by ablaom — APPROVED** [review]
